launcher/ui/themes/CatPack.cpp

```diff
@@ -0,0 +1,113 @@
+#include <qjsonarray.h>
+#include <qjsonobject.h>
+#include <qobject.h>
+#include <QDateTime>
+#include <QDir>
+#include <QFileInfo>
+#include "FileSystem.h"
+#include "Json.h"
+#include "ui/themes/ThemeManager.h"
+
+QString BasicCatPack::path()
+{
+    const QDateTime now = QDateTime::currentDateTime();
+    const QDateTime birthday(QDate(now.date().year(), 11, 30), QTime(0, 0));
+    const QDateTime xmas(QDate(now.date().year(), 12, 25), QTime(0, 0));
+    const QDateTime halloween(QDate(now.date().year(), 10, 31), QTime(0, 0));
+
+    QString cat = QString(":/backgrounds/%1").arg(m_id);
+    if (std::abs(now.daysTo(xmas)) <= 4) {
+        cat += "-xmas";
+    } else if (std::abs(now.daysTo(halloween)) <= 4) {
+        cat += "-spooky";
+    } else if (std::abs(now.daysTo(birthday)) <= 12) {
+        cat += "-bday";
+    }
+    return cat;
+}
+
+JsonCatPack::JsonCatPack(QFileInfo& manifestInfo) : BasicCatPack(manifestInfo.dir().dirName())
+{
+    QString path = FS::PathCombine("catpacks", m_id);
+
+    if (!FS::ensureFolderPathExists(path)) {
+        themeWarningLog() << "couldn't create folder for catpack!";
+        return;
+    }
+
+    if (manifestInfo.exists() && manifestInfo.isFile()) {
+        try {
+            auto doc = Json::requireDocument(manifestInfo.absoluteFilePath(), "CatPack JSON file");
+            const auto root = doc.object();
+            m_name = Json::requireString(root, "name", "Catpack name");
+            auto id = Json::ensureString(root, "id", "", "Catpack ID");
+            if (!id.isEmpty())
+                m_id = id;
+            m_defaultPath = FS::PathCombine(path, Json::requireString(root, "default", "Deafult Cat"));
+            auto variants = Json::ensureArray(root, "variants", QJsonArray(), "Catpack Variants");
+            for (auto v : variants) {
+                auto variant = Json::ensureObject(v, QJsonObject(), "Cat variant");
+                m_variants << Variant{ FS::PathCombine(path, Json::requireString(variant, "path", "Variant path")),
+                                       date(Json::requireString(variant, "startTime", "Variant startTime")),
+                                       date(Json::requireString(variant, "endTime", "Variant endTime")) };
+            }
+
+        } catch (const Exception& e) {
+            themeWarningLog() << "Couldn't load catpack json: " << e.cause();
+            return;
+        }
+    } else {
+        themeDebugLog() << "No catpack json present.";
+    }
+}
+
+QString JsonCatPack::path()
+{
+    const QDate now = QDate::currentDate();
+    for (auto var : m_variants) {
+        QDate startDate(now.year(), var.startTime.month, var.startTime.day);
+        QDate endDate(now.year(), var.endTime.month, var.endTime.day);
+        if (startDate.daysTo(endDate) < 0)  // in this case end date should be next year
+            endDate = endDate.addYears(1);
+        if (startDate.daysTo(now) >= 0 && now.daysTo(endDate) >= 0)
+            return var.path;
```

Comment:
this is still of questionable validity, in my version i had a
```cpp
if (end < start)
    return now <= end || now >= start;
else
    return now >= start && now <= end;
```
this was in a function that returned `bool` if the date range matched so obiously it won't work here, and maybe the `daysTo` does some special logic that comparisons don't

(sorry for complaining about this specific bit of logic so much)

Reply:
Thank, you for catching that. I fixed it by checking if endDate is before or after now, if is before jump one year into the future for endDate, if not jump one year into the past for startDate.
Do not be sorry, as I really do appreciate the feedback. 
If you find more stuff feel free to address them.